### PR TITLE
core/vm: implement full EOF suite [WiP]

### PR DESCRIPTION
**Pull request by ziogaschr**
This PR fetches the changes done on https://github.com/ethereum/go-ethereum/pull/26133 for EOF suite, which community aims to include at Shangai's fork. 
The upstream's work is not finished yet, so this PR will remain open. 
This PR has been created in order we get feedback on it, have progress on Erigon with regards EOF stuff and test them.

- [x] [EIP-3540: EVM Object Format (EOF) v1](https://eips.ethereum.org/EIPS/eip-3540)
- [x] [EIP-3670: EOF - Code Validation](https://eips.ethereum.org/EIPS/eip-3670)
- [x] [EIP-4200: Static relative jumps](https://eips.ethereum.org/EIPS/eip-4200)
- [x] [EIP-4750: EOF - Functions](https://eips.ethereum.org/EIPS/eip-4750)
- [ ] ~~[EIP-5450: EOF - Stack Validation](https://eips.ethereum.org/EIPS/eip-5450)~~ ([WiP lightclient branch](https://github.com/lightclient/go-ethereum/tree/eof-5450))

We tried to keep the code changes similar to upstream's PR, without changing the code at Erigon side, or bring in more changes than needed from upstream.

At this point tests are ~~passing~~ (after we fix a short glitch with regards Shandong's next PR, which make the test fail) and we are able to sync with Shandong testnet.

Please provide feedback, with changes would you like us do.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -28,6 +28,7 @@ require (
 	github.com/edsrzf/mmap-go v1.1.0
 	github.com/emicklei/dot v1.0.0
 	github.com/emirpasic/gods v1.18.1
+	github.com/ethereum/go-ethereum v1.10.26
```
Comment (AskAlexSharov) [inline go.mod:31]: Great work. 
Let's avoid dependency on `go-ethereum`.

**Review by AskAlexSharov — CHANGES_REQUESTED** [review]

File: go.mod (hunk 1 above)
Comment (ziogaschr) [reply go.mod]: Resolved by https://github.com/ledgerwatch/erigon/pull/6215/commits/52816a73839c6fb37c202907428645c4f95ab7e5

**Comment by ziogaschr** [comment]
@AskAlexSharov [here](https://github.com/ETCCooperative/erigon/pull/4) you can find some comments on places in code that triggered our attention. We tried to stick to Erigon logic. Feel free to provide any feedback and we will catch up. Thanks

**Comment by meowsbits** [comment]
My force-push here drops the nearly-latest two commits that together uselessly add, then remove a configuration condition for Shandong that really belonged over [here](https://github.com/ledgerwatch/erigon/pull/6216/commits/cbd13e65ef127b74c692b82e90e6f9b4a2a448af).

**Review by AskAlexSharov — APPROVED** [review]

File: core/vm/contract.go [hunk 2]
```diff
@@ -109,6 +111,14 @@ func isCodeFromAnalysis(analysis []uint64, udest uint64) bool {
 // isCode returns true if the provided PC location is an actual opcode, as
 // opposed to a data-segment following a PUSHN operation.
 func (c *Contract) isCode(udest uint64) bool {
+	if !c.IsLegacy() {
+		// TODO remove this
+		panic("shouldn't run jumpdest analysis on EOF")
+	}
+	// Do we already have an analysis laying around?
+	if c.analysis != nil {
```
Comment (yperbasis) [inline core/vm/contract.go:119]: This is superfluous. `return isCodeFromAnalysis(analysis, udest)` does the same already.

Comment (ziogaschr) [reply core/vm/contract.go]: Thanks @yperbasis. Nice catch.

Resolved by [e30257c](https://github.com/ledgerwatch/erigon/pull/6215/commits/e30257c8811a5eb5ac8c4a3101e1b2f727e7717c)

**Comment by yperbasis** [comment]
Please merge upstream so that there are no conflicts.

File: tests/init.go [hunk 3]
```diff
@@ -227,7 +227,7 @@ var Forks = map[string]*params.ChainConfig{
 		MergeNetsplitBlock:      big.NewInt(0),
 		TerminalTotalDifficulty: big.NewInt(0),
 	},
-	"ArrowGlacierToMergeAtDiffC0000": {
+	"Shanghai": {
```
Comment (yperbasis) [inline tests/init.go:230]: Please keep `ArrowGlacierToMergeAtDiffC0000` as is and introduce `Shanghai` as a new element (copy `Merge` and add `ShanghaiTime`)

**Review by yperbasis — CHANGES_REQUESTED** [review]

**Comment by yperbasis** [comment]
`tests/testdata` submodule was downgraded from https://github.com/ethereum/tests/commit/9e058e565e664ae2ba93f9a25c4beb105df07480 to https://github.com/ethereum/tests/commit/c76c5d274e271e039b0b11e9372f0c1143dd6c30. This is probably by mistake and should be reverted.

File: tests/init.go (hunk 3 above)
Comment (ziogaschr) [reply tests/init.go:245]: Resolved by [581f18f](https://github.com/ledgerwatch/erigon/pull/6215/commits/581f18ffb84918667677e8ae1717aaa874ab1325)

File: core/vm/eips.go [hunk 4]
```diff
@@ -197,3 +201,133 @@ func enable3860(jt *JumpTable) {
...
+	// so no need to verify casts.
+	*pc = uint64(int64(*pc)+int64(relativeOffset)) - 1 // pc will also be increased by interpreter loop
+
+	return nil, nil
+}
+
+// opRjumpi implements the RJUMPI opcode
+func opRjumpi(pc *uint64, interpreter *EVMInterpreter, scope *ScopeContext) ([]byte, error) {
+	condition := scope.Stack.Pop()
+	if condition.BitLen() == 0 {
```
Comment (yperbasis) [inline core/vm/eips.go:256]:
```suggestion
	if condition.IsZero() {
```

File: core/vm/eips.go [hunk 5]
```diff
@@ -197,3 +201,133 @@ func enable3860(jt *JumpTable) {
 	jt[CREATE].dynamicGas = gasCreateEip3860
 	jt[CREATE2].dynamicGas = gasCreate2Eip3860
 }
+
+func enable3540(jt *JumpTable) {
+	// Do nothing.
+}
+
+func enable3670(jt *JumpTable) {
+	// Do nothing.
```
Comment (yperbasis) [inline core/vm/eips.go:210]: Per EIP-3670 `CALLCODE` & `SELFDESTRUCT` should be marked as `legacyOnly`.

Comment (ziogaschr) [reply core/vm/eips.go]: Resolved by [6fbc90f](https://github.com/ledgerwatch/erigon/pull/6215/commits/6fbc90f51c33e283ada243f847e7f52870a31db5)

**Comment by ziogaschr** [comment]
@yperbasis I will try to work on the remaining stuff at original go-ethereum PR or fetch any changes they do, whichever comes first.
Maybe you can focus on other stuff if you prefer.

**Comment by yperbasis** [comment]
Closing in favour of PR #6382.

**Pull request closed without merging**